File: core/src/main/java/xtdb/trie/HashTrie.java
package xtdb.trie;

import org.apache.arrow.memory.util.ArrowBufPointer;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Stream;

public interface HashTrie<N extends HashTrie.Node<N>> {

    int LEVEL_BITS = 2;
    int LEVEL_WIDTH = 1 << LEVEL_BITS;
    int LEVEL_MASK = LEVEL_WIDTH - 1;

    Node<N> rootNode();

    default List<? extends Node<N>> leaves() {
        return rootNode().leaves();
    }

    interface Node<N extends Node<N>> {
        byte[] path();

        N[] children();

        default Stream<? extends HashTrie.Node<N>> leafStream() {
            var children = children();
            return children == null
                    ? Stream.of(this)
                    : Arrays.stream(children).flatMap(child -> child == null ? null : child.leafStream());
        }

        default List<? extends Node<N>> leaves() {
            return leafStream().toList();
        }
    }

    static byte bucketFor(ArrowBufPointer pointer, int level) {
        int bitIdx = level * LEVEL_BITS;
        int byteIdx = bitIdx / Byte.SIZE;
        int bitOffset = bitIdx % Byte.SIZE;

        byte b = pointer.getBuf().getByte(pointer.getOffset() + byteIdx);
        return (byte) ((b >>> ((Byte.SIZE - LEVEL_BITS) - bitOffset)) & LEVEL_MASK);
    }

    static int compareToPath(ArrowBufPointer pointer, byte[] path) {
        for (int level = 0; level < path.length; level++) {
            var cmp = Integer.compare(bucketFor(pointer, level), path[level]);
            if (cmp != 0) return cmp;
        }

        return 0;
    }
}

File: core/src/main/kotlin/xtdb/trie/HashTrie.kt
package xtdb.trie

import xtdb.trie.HashTrie.Node
import java.util.stream.Stream

internal fun conjPath(path: ByteArray, idx: Byte): ByteArray {
    val currentPathLength = path.size
    val childPath = ByteArray(currentPathLength + 1)
    System.arraycopy(path, 0, childPath, 0, currentPathLength)
    childPath[currentPathLength] = idx
    return childPath
}

@Suppress("UNCHECKED_CAST")
private fun <L> Node<L>.leafStream(): Stream<L> =
    hashChildren?.let { it.stream().flatMap { child -> child?.leafStream() } } ?: Stream.of(this as L)

interface HashTrie<L> {
    val rootNode: Node<L>?

    val leaves get() = rootNode?.leaves ?: emptyList()

    interface Node<L> {
        val path: ByteArray

        val hashChildren: List<Node<L>?>?

        val leaves: List<L> get() = leafStream().toList()
    }
}
